### PR TITLE
nezuko/round7: Lion+cosine T_max=24 nocompile (schedule sweep midpoint)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Single-lever experiment: add a cosine LR schedule with **T_max=24** to the SOTA Lion configuration (PR #50). Everything else is identical to the PR #50 baseline.

**Motivation:** Askeladd's PR #57 is testing T_max=16 simultaneously. PR #57 reached val 10.13 at ep9, confirming cosine schedule is stable with Lion uncompiled. This PR provides the **T_max sweep midpoint** — we need to know whether T_max=16 decays too aggressively before the model is ready.

### Why T_max=24?

At the 270-minute training budget (approximately ep9), different T_max values leave very different fractions of peak LR remaining:

| T_max | LR at ep9 | % of peak | Assessment |
|-------|-----------|-----------|------------|
| 16    | ~2.0e-5   | ~40%      | Aggressive decay — may cut LR too early |
| **24** | **~3.5e-5** | **~70%** | **Sweet spot — meaningful decay, substantial LR** |
| 50    | ~4.6e-5   | ~92%      | Barely any decay — schedule has minimal effect |

T_max=24 finds the balance: the model still has substantial learning rate in the final epochs, but the schedule has enough effect to help settle into a better minimum.

### Relation to PR #50 SOTA

PR #50 (vanilla Lion, no schedule) had val 10.08 still descending at budget cutoff — suggesting more headroom exists. A cosine schedule may enable better convergence within the same epoch budget.

---

## Current SOTA Baseline

| Metric | Value | PR |
|--------|-------|-----|
| test_abupt (primary) | **11.208** | PR #50 |
| val metric (best) | ~10.08 | PR #50 (still descending at cutoff) |

Config: Lion uncompiled, lr=5e-5, wd=5e-4, NO schedule, 4L/512d/8h/128slices, ema-decay=0.9995, batch=4, surface/volume points=65536.

---

## Reproduce Command

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent nezuko \
  --wandb-name "nezuko/round7-lion-cosine-tmax24-rank0" \
  --wandb-group "nezuko-round7-lion-cosine-tmax" \
  --optimizer lion --lion-beta1 0.9 --lion-beta2 0.99 \
  --lr 5e-5 --weight-decay 5e-4 \
  --no-compile-model \
  --lr-cosine-t-max 24 \
  --volume-loss-weight 2.0 --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9995 \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms
```

**Single change from PR #50 baseline:** `--lr-cosine-t-max 24` added. All other flags identical.

---

## Context

- **Askeladd PR #57** is testing T_max=16 simultaneously. Compare results once both complete.
- This experiment is the T_max sweep midpoint between T_max=16 (too aggressive?) and T_max=50 (no schedule benefit).
- If T_max=24 beats T_max=16, try T_max=32 or T_max=40 next round.
- If T_max=16 beats T_max=24, the sweet spot is in the shorter range (T_max=12 could be worth testing).

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #93 because ADVISOR: Closing — Lion+cosine T_max=24 nocompile **regresses by +2.8%**.

| Metric | Result | vs SOTA (#50, 11.208) |
|---|---:|---:|
| **test_abupt** | **11.524** | +2.8% |
| best val | 10.301 (ep9) | +2.2% (vs 10.083) |
| W&B run | `ooho1daw` | rt=285m |

Trajectory ran cleanly to completion (9 epochs), but tracked +2-5% behind vanilla Lion uncompiled at every epoch:

| ep | nezuko T=24 | vanilla SOTA |
|---|---:|---:|
| 1 | 74.97 | 80.7 |
| 5 | 14.59 | **14.25** |
| 6 | 12.66 | **12.29** |
| 7 | 11.42 | **11.11** |
| 8 | 10.64 | **10.38** |
| 9 | 10.30 | **10.08** |

**Verdict: Cosine T_max=24 is closed-door on Lion uncompiled.** Combined with edward #51 T_max=16 (test wash) and now T_max=24 (regression), the schedule sweep on Lion uncompiled is exhausted. Next direction: LR sweep (you'll be reassigned to lr=3e-5 to complement alphonse #112 lr=1e-4).

Why it didn't work: vanilla Lion's constant LR is already a near-optimal warm-up-free schedule for this 9-epoch budget. Adding decay early eats useful learning rate too soon..

**Comment by morganmcg1** [comment]
STUDENT nezuko:

## Results — Lion + cosine LR T_max=24 (uncompiled, schedule sweep midpoint)

**Disposition: Negative result.** test_abupt **11.525** vs SOTA 11.208 (**+2.83%**, outside tie band). Both PR #57 (T_max=16, +0.19%) and this PR (T_max=24, +2.83%) confirm cosine LR does **not** help vanilla Lion uncompiled in the 9-epoch / 270-min budget.

### `test_primary/*` — best-val EMA checkpoint reload (n=50 cases)

| Metric | This run (`ooho1daw`) | SOTA PR #50 (`g2n4fyta`) | Δ pp | Δ % | PR #57 T_max=16 |
|---|---:|---:|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` (primary) | **11.525** | 11.208 | +0.317 | +2.83% | 11.229 |
| `surface_pressure_rel_l2_pct` | 6.420 | 6.193 | +0.227 | +3.66% | 6.285 |
| `wall_shear_rel_l2_pct` | 11.522 | 11.199 | +0.323 | +2.88% | 11.278 |
| `volume_pressure_rel_l2_pct` | 13.031 | 12.726 | +0.305 | +2.40% | 12.611 |
| `wall_shear_x_rel_l2_pct` (tau_x) | 9.802 | 9.512 | +0.290 | +3.05% | 9.656 |
| `wall_shear_y_rel_l2_pct` (tau_y) | 13.927 | 13.592 | +0.335 | +2.46% | 13.572 |
| `wall_shear_z_rel_l2_pct` (tau_z) | 14.444 | 14.017 | +0.427 | +3.04% | 14.019 |

Reference (AB-UPT paper, smaller-norm training, fundamentally different setup): surface 3.82, wall_shear 7.29, volume 6.08.

### `val_primary/*` (best-val checkpoint @ epoch 9, n=34)

| Metric | Value |
|---|---:|
| `abupt_axis_mean_rel_l2_pct` | **10.301** |
| `surface_pressure_mae` | 0.01824 |
| `volume_pressure_mae` | 10.792 |
| `wall_shear_mae` | 0.10819 |
| `surface_pressure_rel_l2_pct` | 6.649 |
| `wall_shear_rel_l2_pct` | 11.595 |
| `volume_pressure_rel_l2_pct` | 5.996 |

Best val 10.301 at epoch 9 (last reached) vs PR #50 best val 10.08 (also at last epoch) and PR #57 (T_max=16) best val 10.13.

### Per-epoch val_primary trajectory

| Epoch | abupt_pct | train/lr | LR % of peak |
|---:|---:|---:|---:|
| 1 | 74.97 | 4.98e-5 | 99.7% |
| 2 | 45.44 | 4.92e-5 | 98.4% |
| 3 | 25.70 | 4.81e-5 | 96.2% |
| 4 | 18.07 | 4.67e-5 | 93.4% |
| 5 | 14.59 | 4.49e-5 | 89.8% |
| 6 | 12.66 | 4.28e-5 | 85.6% |
| 7 | 11.42 | 4.04e-5 | 80.8% |
| 8 | 10.64 | 3.78e-5 | 75.6% |
| 9 | **10.30** | 3.49e-5 | 69.8% |

Final LR matches the analytical cosine schedule prediction (PR body said ~3.5e-5/~70%; got 3.49e-5/69.8%) — schedule fired correctly.

Slope analysis (W&B `val/slope/...`): val_primary_abupt was still descending at **−0.199 pp / 1k steps** at end of run. Model was nowhere near refinement phase.

### Run details

- **W&B run (rank 0):** [`ooho1daw`](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/ooho1daw) · group `nezuko-round7-lion-cosine-tmax`
- **Total runtime:** 270.76 min train (hit the 270-min cap exactly) + ~16 min test phase = ~287 min total. Within 360-min env timeout.
- **Peak GPU memory:** ~72.3 GB / 97 GB per H100 (~75% utilization), same envelope as PR #50 baseline
- **Epochs reached:** 9 train / 9 val (matches PR #50, PR #57)
- **No NaNs/divergence** — clean training curve throughout, no grad clipping triggered after warmup
- **train/epoch_loss:** 0.03499 (final) — descending throughout
- **best_checkpoint/source:** ema, valid_primary=1, best_epoch=9

### Exact command used

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent nezuko \
  --wandb-name "nezuko/round7-lion-cosine-tmax24-rank0" \
  --wandb-group "nezuko-round7-lion-cosine-tmax" \
  --optimizer lion --lion-beta1 0.9 --lion-beta2 0.99 \
  --lr 5e-5 --weight-decay 5e-4 \
  --no-compile-model \
  --lr-cosine-t-max 24 \
  --volume-loss-weight 2.0 --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9995 \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms
```

(Identical to PR body; no deviations.)

## What happened

**Cosine T_max=24 underperformed both no-schedule (PR #50) and the more aggressive T_max=16 (PR #57).** Three observations:

1. **All seven test axes regressed by 2.4–3.7%.** This isn't single-axis noise — it's a uniform regression across the entire output, ruling out axis-specific seed quirks. The +2.83% delta on the primary metric is outside the ±2% tie band the team has been using.

2. **The schedule fired correctly.** Final LR landed at 3.49e-5 vs analytically predicted 3.5e-5; model trained the same wallclock and same epoch count as PR #50/#57. The only difference vs PR #50 was the time-varying LR.

3. **Val was still in exploration phase at cutoff.** The −0.199 pp/1k-steps slope at end-of-run, combined with the still-substantial epoch-on-epoch gains (10.64 → 10.30 in the final epoch), means the model needed *more* effective LR for exploration, not less. T_max=24 quietly stole 30% of the LR budget over 9 epochs without giving the model time to use the smaller LR for refinement.

**Why does T_max=24 specifically lose to T_max=16?** This is unintuitive because askeladd's PR #57 (T_max=16) annealed *more* aggressively (final LR 2.07e-5 vs my 3.49e-5) but only got +0.19% test regression. Two candidate explanations:

- **Run-to-run noise.** With single-seed runs, the spread between PR #50 (11.208), PR #57 (11.229), and my run (11.525) could partially reflect seed/dataloader-shuffle variance. PR #57's val 10.13 vs my 10.30 (+1.7%) is at the higher end of typical run-to-run drift in this codebase.
- **Schedule shape interaction with Lion's sign-update.** Lion moves a fixed step size per parameter scaled by LR. Cosine decay between 100% and ~70% of peak (T_max=24) keeps Lion in a regime where step sizes are moderately reduced but still large enough that the EMA reset on each step keeps generating new noise. Aggressive decay (T_max=16, ~40% peak) forces faster averaging into a smaller-step regime that is at least more internally consistent. T_max=24 may be in a worst-of-both-worlds zone — too much annealing to maintain peak-LR exploration, not enough to compress into a stable refinement regime.

**Verdict on the lever:** Combined PR #50 / #57 / #93 evidence: cosine LR is a **closed lever** for Lion uncompiled in this 9-epoch / 270-min budget. The model never finishes exploration, so any schedule that reduces LR before the model is ready loses. To re-open this lever would require either (a) compile working with Lion to get more epochs (currently diverges, see PR #50 / #57 mid-run notes), or (b) testing T_max=9 with `--lr-min` set to something modest like 1e-5 so the floor is reached but a meaningful LR is preserved.

## Suggested follow-ups

1. **Drop cosine for Lion uncompiled and move on.** This is the team's third Lion+cosine experiment in two days; the lever is exhausted at this budget. Recommend closing both this PR and the cosine-on-Lion direction unless compile is fixed.
2. **Try a mild constant-then-decay schedule (CALR or step at ep7).** Keep LR=5e-5 for the first 7-8 epochs of exploration, then drop sharply to ~5e-6 for the last 1-2 epochs. This separates exploration from refinement more cleanly than cosine, and matches the empirical observation that val improvement decelerates only in the last 2 epochs.
3. **Re-test cosine after we get more epochs.** Once any path to >12 epochs/budget exists (compile-fix, larger batch, fewer val cases, smaller model, etc.), cosine becomes worth re-testing: with 14+ epochs, T_max=14 with full decay to lr_min=1e-6 would actually let refinement happen.
4. **Investigate Lion + compile divergence as a higher-leverage problem.** Both nezuko and askeladd have hit the same `aten._foreach_lerp_*` chaos at epoch 4. Fixing this unlocks +30–40% throughput for Lion → 12+ epochs in 270 min → cosine becomes viable again.
